### PR TITLE
build: bump test dependencies

### PR DESCRIPTION
**Pull request by peterpeterparker**
# Motivation

Upgrade the test dependencies.

# Changes

- Bump vitest and related dependencies
- Adapt `vitest.config` for breaking changes
  - Add `svelteTesting` plugin
  - Move `useAtomics` config

### CODE REVIEW

**Comment by peterpeterparker** [comment]
@dskloetd big PR but, there is also a big gap between vitest `0.34.6` and `2.1.1`. Not sure if there was ways to break things down.

Thanks a ton @mstrasinskis for the help!

File: frontend/src/tests/lib/services/worker-transactions.services.spec.ts [hunk 1]
```diff
@@ -4,22 +4,21 @@ import type { PostMessageDataRequestTransactions } from "$lib/types/post-message
 import { mockSnsMainAccount } from "$tests/mocks/sns-accounts.mock";
 import { indexCanisterIdMock } from "$tests/mocks/sns.api.mock";
 
-describe("initTransactionsWorker", () => {
-  let spyPostMessage;
+let spyPostMessage;
+vi.doMock("$lib/workers/transactions.worker?worker", () => ({
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/worker-transactions.services.spec.ts:8]: The PR description says

> Replace doMock with mock in tests that use Workers, as doMock doesn’t prevent worker files from being initialised, which triggers the ReferenceError: Worker is not defined error.

But this is still `doMock`. Should this be `mock`?

File: frontend/src/tests/lib/services/worker-transactions.services.spec.ts [hunk 2]
```diff
@@ -4,22 +4,21 @@ import type { PostMessageDataRequestTransactions } from "$lib/types/post-message
...
+  default: class TransactionsWorker {
+    postMessage(data: {
+      msg: "nnsStartTransactionsTimer";
+      data: PostMessageDataRequestTransactions;
+    }) {
+      spyPostMessage(data);
+    }
+  },
+}));
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/worker-transactions.services.spec.ts:17]: Why did the mock need to be moved outside of `beforeEach`?

Can it also be moved without updating the dependencies or does this require updated dependencies?

File: frontend/src/tests/lib/services/sns.services.spec.ts [hunk 3]
```diff
@@ -282,7 +282,7 @@ describe("sns-services", () => {
   });
 
   describe("loadSnsSwapCommitment", () => {
-    let queryCommitmentSpy: SpyInstance;
+    let queryCommitmentSpy: MockInstance;
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/sns.services.spec.ts:285]: Is `MockInstance` a new type or could this change be moved before updating the dependencies?

File: frontend/src/tests/lib/services/public/proposals.services.spec.ts [hunk 4]
```diff
@@ -224,7 +224,7 @@ describe("proposals-services", () => {
 
         await loadProposal({
           proposalId: 0n,
-          setProposal: vi.fn,
+          setProposal: vi.fn(),
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/public/proposals.services.spec.ts:227]: This looks like a bug regardless of dependencies.
Why did this not cause any problems before and why did it start causing problems now?

File: frontend/src/tests/lib/routes/Wallet.spec.ts [hunk 5]
```diff
@@ -161,7 +161,7 @@ describe("Wallet", () => {
   });
 
   describe("sns context", () => {
-    beforeAll(() => {
+    beforeEach(() => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/routes/Wallet.spec.ts:164]: This is a good change but does not seem to be related to updating the dependencies.

File: frontend/src/tests/lib/routes/Settings.spec.ts [hunk 6]
```diff
@@ -35,27 +36,30 @@ describe("Settings", () => {
...
 
-  it("should render a dynamic expired session time", () => {
+  it("should render a dynamic expired session time", async () => {
     const { getByTestId, rerender } = render(Settings);
 
     const element = getByTestId("session-duration");
     expect(element?.textContent.trim() ?? "").toEqual("");
 
     authRemainingTimeStore.set(250000);
+    await runResolvedPromises();
```
Comment (dskloetd) [inline frontend/src/tests/lib/routes/Settings.spec.ts:46]: Why was this not necessary before and why is it necessary now?

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts [hunk 7]
```diff
@@ -160,6 +160,14 @@ describe("NnsProposalDetail", () => {
                 }
           )
         );
+      vi.spyOn(governanceApi, "queryNeurons").mockImplementation(
+        ({ certified }) =>
+          certified
+            ? Promise.resolve(testNeurons)
+            : // Throw just to be sure that only certified version is used.
+              // Otherwise neuronsStoreReady is never true: https://github.com/dfinity/nns-dapp/blob/07a9e3e53df469e63798fdf9c7fc0317eeda79c1/frontend/src/lib/components/proposal-detail/VotingCard/NnsVotingCard.svelte#L126
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts:168]: This link points to a blank line.

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts [hunk 8]
```diff
@@ -160,6 +160,14 @@ describe("NnsProposalDetail", () => {
                 }
           )
         );
+      vi.spyOn(governanceApi, "queryNeurons").mockImplementation(
+        ({ certified }) =>
+          certified
+            ? Promise.resolve(testNeurons)
+            : // Throw just to be sure that only certified version is used.
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts:167]: I'm not following this.
Is it bad if `queryNeurons` is called with `certified: false` during the test? Or is it expected?
Why does responding to such a call result in `neuronsStoreReady` never being true?

And why was this not an issue before updating the dependencies?

File: frontend/src/tests/lib/components/ui/Hash.spec.ts [hunk 9]
```diff
@@ -32,7 +32,8 @@ describe("Hash", () => {
       props: { text: identifier, testId, id: identifier },
     });
 
-    const tooltipElement = container.querySelector("[role='tooltip']");
+    const body = container.ownerDocument.body;
+    const tooltipElement = body.querySelector("[role='tooltip']");
```
Comment (dskloetd) [inline frontend/src/tests/lib/components/ui/Hash.spec.ts:36]: Why was this not necessary before?

File: frontend/src/tests/lib/routes/Wallet.spec.ts (hunk 5 above)
Comment (dskloetd) [reply frontend/src/tests/lib/routes/Wallet.spec.ts:164]: I looked into this a bit more.
After the deps update, the first test started redirecting to `/tokens`, leaving the second test to start at the wrong page. This can indeed be fixed by changing `beforeAll` to `beforeEach`.
But why did this start happening?
I found out that the first test is redirecting [here](https://github.com/dfinity/nns-dapp/blob/18b61e78de571d13708b90746eb85f981a08a989/frontend/src/lib/components/accounts/IcrcWalletPage.svelte#L101) because a few lines earlier
```
if (nonNullish($selectedAccountStore.account)) {
```
fails to be true, which would return from the function before the page get redirected.
This happens even though the store does actually contain a valid `account`.
If on that line you replace `$selectedAccountStore.account` with `get(selectedAccountStore)` then the test does not redirect and the test passes (without fixing the `beforeAll`).
Using `get` instead of `$` in places that don't rely on dynamic updates might be a good idea regardless but I didn't know that it could actually result in different values.

Comment (mstrasinskis) [reply frontend/src/tests/lib/routes/Wallet.spec.ts:164]: Good catch!

File: frontend/src/tests/routes/app/home/home-page.spec.ts [hunk 10]
```diff
@@ -1,16 +1,14 @@
...
 import { render } from "@testing-library/svelte";
-import { mock } from "vitest-mock-extended";
 
 describe("Home page", () => {
   beforeEach(() => {
     overrideFeatureFlagsStore.reset();
 
-    // TODO: agent mocked because some calls to global.fetch were exposed when we migrated to agent-js v0.20.2
-    vi.spyOn(agent, "createAgent").mockResolvedValue(mock<HttpAgent>());
+    vi.spyOn(icrcLedgerApi, "queryIcrcToken").mockResolvedValue(mockCkBTCToken);
```
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply frontend/src/tests/routes/app/home/home-page.spec.ts]: Addressed in a separate PR.

File: frontend/src/tests/lib/services/worker-transactions.services.spec.ts (hunk 1 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/worker-transactions.services.spec.ts]: Addressed in a separate PR.

File: frontend/src/tests/lib/services/worker-transactions.services.spec.ts (hunk 2 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/worker-transactions.services.spec.ts]: Addressed in a separate PR.

File: frontend/src/tests/lib/services/sns.services.spec.ts (hunk 3 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/sns.services.spec.ts]: Addressed in a separate PR.

File: frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts (hunk 7 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/pages/NnsProposalDetail.spec.ts]: Addressed in a separate PR.

File: frontend/src/tests/lib/routes/Wallet.spec.ts (hunk 5 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/routes/Wallet.spec.ts:164]: Addressed in a separate PR.

File: frontend/src/tests/lib/components/ui/Hash.spec.ts (hunk 9 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/components/ui/Hash.spec.ts]: Addressed in a separate PR.

File: frontend/vitest.config.ts [hunk 11]
```diff
@@ -1,11 +1,12 @@
 import { sveltekit } from "@sveltejs/kit/vite";
+import { svelteTesting } from "@testing-library/svelte/vite";
 import { resolve } from "path";
 import type { UserConfig } from "vite";
 import { configDefaults, defineConfig } from "vitest/config";
 
 export default defineConfig(
   ({ mode }: UserConfig): UserConfig => ({
-    plugins: [sveltekit()],
+    plugins: [sveltekit(), svelteTesting()],
```
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply frontend/vitest.config.ts:9]: @peterpeterparker, do you remember the intent behind adding it?

[the remark this answers is not in the document]
Comment (dskloetd) [reply frontend/vitest.config.ts:9]: I did some testing and without this things like `mockImplementation` and `mockReturnValue` didn't seem to exist anymore.

[the remark this answers is not in the document]
Comment (peterpeterparker) [reply frontend/vitest.config.ts:9]: I think it was previously bundled within the testing library but has become necessary in latest version.

Adding `svelteTesting` is part of the Testing Library setup: https://testing-library.com/docs/svelte-testing-library/setup/

File: CHANGELOG-Nns-Dapp-unreleased.md [hunk 12]
```diff
@@ -20,6 +20,7 @@ proposal is successful, the changes it released will be moved from this file to
 
 #### Changed
 
+* Upgrade frontend dev dependencies (`SvelteKit v2`, `vite`, `vitest`, etc.)
```
Comment (dskloetd) [inline CHANGELOG-Nns-Dapp-unreleased.md:23]: The `"@sveltejs/kit"` dependency is unchanged at `"^2.5.28"`.
What does "SvelteKit v2" refer to?

Maybe a change log entry is not necessary but if we think some of the version changes are so big it's worth mentioning, let's be accurate.

**Review by dskloetd — APPROVED** [review]

File: CHANGELOG-Nns-Dapp-unreleased.md (hunk 12 above)
Comment (mstrasinskis) [reply CHANGELOG-Nns-Dapp-unreleased.md]: Removed.

File: frontend/vitest.config.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply frontend/vitest.config.ts:9]: Thanks for the explanation.

File: frontend/src/tests/lib/services/public/proposals.services.spec.ts (hunk 4 above)
Comment (mstrasinskis) [reply frontend/src/tests/lib/services/public/proposals.services.spec.ts]: Addressed in a separate pr.
